packages/venia-concept/src/components/ProductFullDetail/ProductFullDetail.js

```diff
@@ -1,254 +1,241 @@
 import { arrayOf, bool, func, number, shape, string } from 'prop-types';
 import { Form } from 'informed';
 import { Price } from '@magento/peregrine';
 
-import classify from 'src/classify';
+import defaultClasses from './productFullDetail.css';
+import { mergeClasses } from 'src/classify';
+
 import Button from 'src/components/Button';
 import { loadingIndicator } from 'src/components/LoadingIndicator';
 import Carousel from 'src/components/ProductImageCarousel';
 import Quantity from 'src/components/ProductQuantity';
 import RichText from 'src/components/RichText';
-import defaultClasses from './productFullDetail.css';
+
 import appendOptionsToPayload from 'src/util/appendOptionsToPayload';
 import findMatchingVariant from 'src/util/findMatchingProductVariant';
 import isProductConfigurable from 'src/util/isProductConfigurable';
 
 const Options = React.lazy(() => import('../ProductOptions'));
 
-class ProductFullDetail extends Component {
-    static propTypes = {
-        classes: shape({
-            cartActions: string,
-            description: string,
-            descriptionTitle: string,
-            details: string,
-            detailsTitle: string,
-            imageCarousel: string,
-            options: string,
-            productName: string,
-            productPrice: string,
-            quantity: string,
-            quantityTitle: string,
-            root: string,
-            title: string
-        }),
-        product: shape({
-            __typename: string,
-            id: number,
-            sku: string.isRequired,
-            price: shape({
-                regularPrice: shape({
-                    amount: shape({
-                        currency: string.isRequired,
-                        value: number.isRequired
-                    })
-                }).isRequired
-            }).isRequired,
-            media_gallery_entries: arrayOf(
-                shape({
-                    label: string,
-                    position: number,
-                    disabled: bool,
-                    file: string.isRequired
-                })
-            ),
-            description: string
-        }).isRequired,
-        addToCart: func.isRequired
-    };
-
-    static getDerivedStateFromProps(props, state) {
-        const { configurable_options } = props.product;
-        const optionCodes = new Map(state.optionCodes);
+const INITIAL_OPTION_CODES = new Map();
+const INITIAL_OPTION_SELECTIONS = new Map();
+const INITIAL_QUANTITY = 1;
 
-        // if this is a simple product, do nothing
-        if (!isProductConfigurable(props.product)) {
-            return null;
-        }
-
-        // otherwise, cache attribute codes to avoid lookup cost later
-        for (const option of configurable_options) {
-            optionCodes.set(option.attribute_id, option.attribute_code);
-        }
-
-        return { optionCodes };
+const deriveOptionCodesFromProduct = product => {
```

Comment:
I think this is really great feedback in the general case but I think the `ProductFullDetail` component is an exception to the rule. Here's why:

> See how initially state.optionCodes will be an empty map but any time props are provided we update the state with more and more option codes.

Great call but we don't actually want to do this. Playing that out, the list of option codes would just get bigger and bigger based on how many different `product`s got sent in as props, and the list would lose its meaning. We really only care about *this* product's option codes.

For an illustrative example, we determine if all options have been selected by comparing the number of selections (`optionSelections`) to the number of options available (`optionCodes`). We don't want the number of options available to include other products' options.

I think the reason this "bug" was never uncovered was because we never actually re-use this component and pass a new `product` prop to it. When users navigate away from the product detail page, `ProductFullDetail` is unmounted. When they navigate back, a new instance is mounted (with a different `product` prop).

Reply:
There's never a time in which the PDP is re-used? Not even when something changes or a selection is chosen?

Reply:
Those are all state changes, not new props being passed in.

Reply:
I am incorrect, `getDerivedStateFromProps` does get called every time a selection is chosen.

But the props never change and therefore the return value never changes. Check out these console logs:

![Screen Shot 2019-05-22 at 12 09 50 PM](https://user-images.githubusercontent.com/13182778/58194268-b37aa080-7c8a-11e9-9bc3-be996b87af19.png)

When I first load the page `optionCodes` is an empty `Map`. Every time I make a selection `getDerivedStateFromProps` runs but nothing is updated.

When I load up the PDP with a different product the `optionCodes` starts again as an empty `Map`.

Therefore I'm confident we aren't losing any functionality here.

Thanks for your patience on this one!

Reply:
> But the props never change and therefore the return value never changes. 

Sounds perfect for a memoized callback XD